File: gpi back/gpi-backend/src/main/java/com/gpi/gpi_backend/service/BanqueService.java
package com.gpi.gpi_backend.service;

import com.gpi.gpi_backend.dto.BanqueDTO;
import com.gpi.gpi_backend.model.Banque;
import com.gpi.gpi_backend.repository.BanqueRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class BanqueService {

    private final BanqueRepository repository;

    public List<BanqueDTO> getAll() {
        return repository.findAll()
                .stream().map(this::toDTO)
                .collect(Collectors.toList());
    }

    public BanqueDTO getById(Long id) {
        return toDTO(repository.findById(id)
                .orElseThrow(() -> new RuntimeException("Banque non trouvée : " + id)));
    }

    public BanqueDTO create(BanqueDTO dto) {
        if (repository.existsByBic(dto.getBic())) {
            throw new RuntimeException("BIC déjà existant : " + dto.getBic());
        }
        return toDTO(repository.save(toEntity(dto)));
    }

    public BanqueDTO update(Long id, BanqueDTO dto) {
        Banque b = repository.findById(id)
                .orElseThrow(() -> new RuntimeException("Banque non trouvée : " + id));
        b.setBic(dto.getBic());
        b.setNomBanque(dto.getNomBanque());
        b.setPaysCode(dto.getPaysCode());
        b.setPaysNom(dto.getPaysNom());
        b.setFlagUrl(dto.getFlagUrl());
        b.setTypeBanque(dto.getTypeBanque());
        b.setDevise(dto.getDevise());
        b.setCutOff(dto.getCutOff());
        b.setFuseauHoraire(dto.getFuseauHoraire());
        return toDTO(repository.save(b));
    }

    public void delete(Long id) {
        if (!repository.existsById(id)) {
            throw new RuntimeException("Banque non trouvée : " + id);
        }
        repository.deleteById(id);
    }

    private BanqueDTO toDTO(Banque b) {
        return BanqueDTO.builder()
                .id(b.getId())
                .bic(b.getBic())
                .nomBanque(b.getNomBanque())
                .paysCode(b.getPaysCode())
                .paysNom(b.getPaysNom())
                .flagUrl(b.getFlagUrl())
                .typeBanque(b.getTypeBanque())
                .devise(b.getDevise())
                .cutOff(b.getCutOff())
                .fuseauHoraire(b.getFuseauHoraire())
                .build();
    }

    private Banque toEntity(BanqueDTO dto) {
        return Banque.builder()
                .bic(dto.getBic())
                .nomBanque(dto.getNomBanque())
                .paysCode(dto.getPaysCode())
                .paysNom(dto.getPaysNom())
                .flagUrl(dto.getFlagUrl())
                .typeBanque(dto.getTypeBanque())
                .devise(dto.getDevise())
                .cutOff(dto.getCutOff())
                .fuseauHoraire(dto.getFuseauHoraire())
                .build();
    }
}

File: gpi front/src/app/services/banque.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Bank {
  id?: number;
  bic: string;
  nomBanque: string;
  paysCode: string;
  paysNom: string;
  flagUrl: string;
  typeBanque: string;
  devise: string;
  cutOff: string;
  fuseauHoraire: string;
}

@Injectable({ providedIn: 'root' })
export class BanqueService {
  private api = 'http://localhost:8080/api/banques';

  constructor(private http: HttpClient) {}

  getAll(): Observable<Bank[]> {
    return this.http.get<Bank[]>(this.api);
  }

  create(bank: Bank): Observable<Bank> {
    return this.http.post<Bank>(this.api, bank);
  }

  update(id: number, bank: Bank): Observable<Bank> {
    return this.http.put<Bank>(`${this.api}/${id}`, bank);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.api}/${id}`);
  }
}